mayastor/src/bdev/nexus/nexus_io_subsystem.rs

```diff
@@ -0,0 +1,182 @@
+    /// handle internal events and which is a protocol feature.
+    /// In case concurrent pause requests take place, the other callers
+    /// will wait till the nexus is resumed and will continue execution
+    /// with the nexus paused once they are awakened via resume().
+    /// Note: in order to handle concurrent pauses properly, this function must
+    /// be called only from the master core.
+    pub async fn suspend(&mut self) -> Result<(), NexusError> {
+        assert_eq!(Cores::current(), Cores::first());
+
+        trace!(?self.name, "pausing nexus I/O");
+
+        loop {
+            let state = self.pause_state.compare_exchange(
+                NexusPauseState::Unpaused,
+                NexusPauseState::Pausing,
+            );
+
+            match state {
+                Ok(NexusPauseState::Unpaused) => {
+                    // Pause subsystem. The only acceptable counter transition
+                    // is: 0 -> 1.
+                    assert_eq!(
+                        self.pause_cnt.fetch_add(1, Ordering::SeqCst),
+                        0,
+                        "Corrupted subsystem pause counter"
+                    );
+
+                    if let Some(Protocol::Nvmf) = self.bdev.shared() {
+                        if let Some(subsystem) =
+                            NvmfSubsystem::nqn_lookup(&self.name)
+                        {
+                            trace!(nexus=%self.name, nqn=%subsystem.get_nqn(), "pausing subsystem");
+                            subsystem.pause().await.unwrap();
+                            trace!(nexus=%self.name, nqn=%subsystem.get_nqn(), "subsystem paused");
+                        }
+                    }
+
+                    // Mark subsystem as paused after it has been paused.
+                    self.pause_state
+                        .compare_exchange(
+                            NexusPauseState::Pausing,
+                            NexusPauseState::Paused,
+                        )
+                        .expect("Failed to mark subsystem as Paused");
+                    break;
+                }
+                // Subsystem is already paused, increment number of paused.
+                Err(NexusPauseState::Paused) => {
+                    trace!(nexus=%self.name, "nexus is already paused, incrementing refcount");
+                    self.pause_cnt.fetch_add(1, Ordering::SeqCst);
+                    break;
+                }
+                // Wait till the subsystem has completed transition and retry
+                // operation.
+                Err(NexusPauseState::Unpausing)
+                | Err(NexusPauseState::Pausing) => {
+                    trace!(nexus=%self.name, "nexus is in intermediate state, deferring Pause operation");
+                    let (s, r) = oneshot::channel::<i32>();
+                    self.pause_waiters.push_back(s);
+                    r.await.unwrap();
```

Comment:
what if the nexus is destroyed while a Pause is in progress?

Reply:
Upon nexus removal all its child devices are destroyed, all its I/O is completed.  Since this code path is part of I/O path, it's also be completed before proceeding with nexus removal.